### PR TITLE
generate seccomp profile convert type

### PR DESCRIPTION
**Pull request by jessfraz**
fixes the problem noted by @runcom  here https://github.com/docker/docker/pull/20106#issuecomment-185254234

### CODE REVIEW

**Comment by runcom** [comment]
can we test load the default profile also here? https://github.com/docker/docker/blob/master/profiles/seccomp/seccomp_test.go#L10? I was going to work on this (but you made it first) and I wrote the basic test which was clearly failing :)

**Comment by runcom** [comment]
Apart from the unit test, this is working fine for me:
```sh
$ docker run --rm --security-opt seccomp:`pwd`/profiles/seccomp/default.json -i debian:jessie \
    unshare --map-root-user --user sh -c whoami
unshare: unshare failed: Operation not permitted
```

**Comment by jfrazelle** [comment]
Will update later today

File: profiles/seccomp/generate.go [hunk 1]
```diff
@@ -4,13 +4,138 @@ package main
...
+	"mips64":      types.ArchMIPS64,
+	"mips64n32":   types.ArchMIPS64N32,
+	"mipsel":      types.ArchMIPSEL,
+	"mipsel64":    types.ArchMIPSEL64,
+	"mipsel64n32": types.ArchMIPSEL64N32,
+}
+
+// convertOperator converts a Seccomp comparison operator.
+func convertOperator(in configs.Operator) (types.Operator, error) {
+	if op, ok := operators[in]; ok == true {
```
Comment (runcom) [inline profiles/seccomp/generate.go]: can you drop `== true`? also below twice

**Comment by runcom** [comment]
alright :)

LGTM

**Comment by jfrazelle** [comment]
sorry I redid how my original commit was working you should take another look, also thanks :)

**Comment by runcom** [comment]
@jfrazelle is it missing the type conversion for `default.json`? cause tests fail for this (original issue - default.json cannot be used with --security-opt seccomp:default.json)

**Comment by jfrazelle** [comment]
wth, i made the type for the default profile types.seccomp :(, crap I will have to look later thats so weird

**Comment by runcom** [comment]
I think the previous change you did was the right one - use configs.seccomp for execdriver and engine-types types.seccomp for profiles/seccomp/default.json in the repository. My tests against your repo was indeed ensuring that default.json can be used with --security-opt seccomp:... (which right now it's not working but did work with your original patch)

**Comment by jfrazelle** [comment]
my bad... fixed

**Comment by runcom** [comment]
can you re-add this test https://github.com/jfrazelle/docker/pull/10/files#diff-b42881ac05309b4a27d08e860713de7bR20? this should be green

**Comment by runcom** [comment]
> ob: Docker-PRs-userns FAILED:
> 
> ---
> od  0.853s
> PASS: docker_cli_run_unix_test.go:807: DockerSuite.TestRunSeccompProfileDenyCloneUserns 3.627s
> PASS: docker_cli_run_unix_test.go:713: DockerSuite.TestRunSeccompProfileDenyUnshare 3.128s
> PASS: docker_cli_run_unix_test.go:770: DockerSuite.TestRunSeccompProfileDenyUnshareUserns   2.575s
> SKIP: docker_cli_run_unix_test.go:819: DockerSuite.TestRunSeccompUnconfinedCloneUserns (Test cannot be run when remapping root)
> 
> ----------------------------------------------------------------------
> FAIL: docker_cli_run_unix_test.go:915: DockerSuite.TestRunSeccompWithDefaultProfile
> 
> docker_cli_run_unix_test.go:920:
>     c.Assert(out, checker.Equals, "unshare: unshare failed: Operation not permitted")
> ... obtained string = "/go/src/github.com/docker/docker/bundles/1.11.0-dev/test-integration-cli/../binary/docker: Error response from daemon: Cannot start container ccb197c249cc2816064def45d8c6ca1d1e610f126b100314b2010332a954e110: Decoding seccomp profile failed: json: cannot unmarshal number into Go value of type types.Action.\n"
> ... expected string = "unshare: unshare fa
> ---

this is what I'm saying - the default profile at `profiles/seccomp/default.json` isn't using engine-types (but libcontainer/configs) and so it's not readable by --security-opt seccomp:...

**Comment by jfrazelle** [comment]
Ugh dammit I didn't push sorry I won't be able to update til later

**Comment by jfrazelle** [comment]
It's because generate is still calling loadProfile which changes the default type to the config type but this does work w the default profile being of types.Seccomp which makes the patch complexity less large

**Comment by jfrazelle** [comment]
My brain has already shut off lol, I'll be able to push in a few hours

On Friday, February 19, 2016, Gordon <notifications@github.com> wrote:

> Job: Docker-PRs-experimental FAILED
> <https://jenkins.dockerproject.org/job/Docker-PRs-experimental/15178/console>
> :
>
> ---Suite.TestRunSeccompProfileDenyChmod    1.259sPASS: docker_cli_run_unix_test.go:807: DockerSuite.TestRunSeccompProfileDenyCloneUserns 2.790sPASS: docker_cli_run_unix_test.go:713: DockerSuite.TestRunSeccompProfileDenyUnshare 2.514sPASS: docker_cli_run_unix_test.go:770: DockerSuite.TestRunSeccompProfileDenyUnshareUserns   2.458sPASS: docker_cli_run_unix_test.go:819: DockerSuite.TestRunSeccompUnconfinedCloneUserns  2.469s
> ----------------------------------------------------------------------FAIL: docker_cli_run_unix_test.go:915: DockerSuite.TestRunSeccompWithDefaultProfile
> docker_cli_run_unix_test.go:920:    c.Assert(out, checker.Equals, "unshare: unshare failed: Operation not permitted")... obtained string = "/go/src/github.com/docker/docker/bundles/1.11.0-dev/test-integration-cli/../binary/docker: Error response from daemon: Cannot start container 9189985a0e90224d8f13b9be15ac6949d167bf6c49df9d5ab0861c8f00c6a965: Decoding seccomp profile failed: json: cannot unmarshal number into Go value of type types.Action.\n"... expected string = "unshare: unshare fa---
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/20428#issuecomment-186325981>.
>


-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

**Comment by runcom** [comment]
> My brain has already shut off lol, I'll be able to push in a few hours

ahah don't worry I guess we can wait till next week :)

**Comment by jfrazelle** [comment]
ok pushed

**Comment by runcom** [comment]
Thanks @jfrazelle, LGTM if janky's will ever be happy

**Comment by justincormack** [comment]
Ah this should fix #20678 which means can fix the CI to merge #20634 which would be nice.

**Comment by runcom** [comment]
@justincormack yup, ping @LK4D4 @calavera @cpuguy83

**Comment by thaJeztah** [comment]
Oh! Thought this was merged already, ha

**Comment by justincormack** [comment]
LGTM

**Comment by vdemeester** [comment]
LGTM 🐭

**Comment by cpuguy83** [comment]
LGTM

**Comment by cpuguy83** [comment]
userns issues are known and not caused by this PR, merging.

**Comment by runcom** [comment]
thanks @jfrazelle

**Pull request merged**